File: typescript/problem_63.ts
export function wordInMatrix(matrix: string[][], word: string): boolean {
    const rows = matrix.length, cols = matrix[0].length, n = word.length;
    for (let r = 0; r < rows; r++)
        for (let c = 0; c <= cols - n; c++)
            if ([...word].every((ch, i) => matrix[r][c+i] === ch)) return true;
    for (let c = 0; c < cols; c++)
        for (let r = 0; r <= rows - n; r++)
            if ([...word].every((ch, i) => matrix[r+i][c] === ch)) return true;
    return false;
}

export function problem63(): void {
    console.log("Problem 63:");
    const m = [['F','A','C','I'],['O','B','Q','P'],['A','N','O','B'],['M','A','S','S']];
    if (!wordInMatrix(m, "FOAM")) throw new Error("assertion failed");
    if (!wordInMatrix(m, "MASS")) throw new Error("assertion failed");
    if ( wordInMatrix(m, "NOPE")) throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: rust/problem_63.rs
pub fn word_in_matrix(matrix: &Vec<Vec<char>>, word: &str) -> bool {
    let chars: Vec<char> = word.chars().collect();
    let n = chars.len();
    let rows = matrix.len(); let cols = matrix[0].len();
    for r in 0..rows {
        for c in 0..cols.saturating_sub(n-1) {
            if (0..n).all(|i| matrix[r][c+i] == chars[i]) { return true; }
        }
    }
    for c in 0..cols {
        for r in 0..rows.saturating_sub(n-1) {
            if (0..n).all(|i| matrix[r+i][c] == chars[i]) { return true; }
        }
    }
    false
}

pub fn problem63() {
    println!("Problem 63:");
    let m = vec![
        vec!['F','A','C','I'], vec!['O','B','Q','P'],
        vec!['A','N','O','B'], vec!['M','A','S','S'],
    ];
    assert!( word_in_matrix(&m, "FOAM"));
    assert!( word_in_matrix(&m, "MASS"));
    assert!(!word_in_matrix(&m, "NOPE"));
    println!("All Tests Passed");
}
